datumaro/plugins/synthetic_data/image_generator.py

```diff
@@ -0,0 +1,179 @@
+            pool.starmap(self._generate_image_batch, generation_params)
+
+    def _generate_image_batch(self, params: np.ndarray, weights: np.ndarray, indices: List[int]) -> None:
+        path = osp.dirname(osp.abspath(__file__))
+        proto = osp.join(path, 'colorization_deploy_v2.prototxt')
+        model = osp.join(path, 'colorization_release_v2.caffemodel')
+        npy = osp.join(path, 'pts_in_hull.npy')
+        pts_in_hull = np.load(npy).transpose().reshape(2, 313, 1, 1).astype(np.float32)
+        net = cv.dnn.readNetFromCaffe(proto, model)
+        net.getLayer(net.getLayerId('class8_ab')).blobs = [pts_in_hull]
+        net.getLayer(net.getLayerId('conv8_313_rh')).blobs = [np.full([1, 313], 2.606, np.float32)]
+
+        for i, param, w in zip(indices, params, weights):
+            image = self._generator(param, self._iterations, self._height, self._width, draw_point=False, weight=w)
+            color_image = colorize(image, net)
+            aug_image = augment(self._rng, color_image)
+            cv.imwrite(osp.join(self._output_dir, "{:06d}.png".format(i)), aug_image)
+
+    def _generator(self, params: np.ndarray, iterations: int, height: int, width: int,
+                   draw_point: bool = True, weight: Optional[np.ndarray] = None) -> np.ndarray:
+        ifs_function = IFSFunction(self._rng, prev_x=0.0, prev_y=0.0)
+        for param in params:
+            ifs_function.set_param(param[:6], param[6], weight)
+        ifs_function.calculate(iterations)
+        img = ifs_function.draw(height, width, draw_point)
+        return img
+
+    def _generate_category(self, eps: float, base_h: int = 512, base_w: int = 512) -> np.ndarray:
+        pixels = -1
+        while pixels < self._threshold:
+            param_size = self._rng.randint(2, 7)
+            params = np.zeros((param_size, 7), dtype=np.float32)
+
+            sum_proba = eps
+            for i in range(param_size):
+                a, b, c, d, e, f = [self._rng.uniform(-1.0, 1.0) for _ in range(6)]
+                prob = abs(a * d - b * c)
+                sum_proba += prob
+                params[i] = a, b, c, d, e, f, prob
+            params[:, 6] /= sum_proba
+
+            fracral_img = self._generator(params, self._num_of_points, base_h, base_w)
+            pixels = np.count_nonzero(fracral_img) / (base_h * base_w)
+        return params
+
+    def _initialize_params(self) -> None:
+        default_img_size = 362 * 362
+        points_coeff = max(1, int(np.round(self._height * self._width / default_img_size)))
+        self._num_of_points = 100000 * points_coeff
+
+        if self._count < len(self._weights):
+            self._instances = 1
+            self._categories = 1
+            self._weights = self._weights[:self._count, :]
+        else:
+            self._instances = np.ceil(0.25 * self._count / self._weights.shape[0]).astype(int)
+            self._categories = np.ceil(self._count / (self._instances * self._weights.shape[0])).astype(int)
```

Comment:
Maybe I miss something, but it looks like `categories` are almost (?) always 4.

Reply:
Fixed